astropy/utils/data.py

```diff
@@ -1024,9 +1034,91 @@ def connect_ftp(self, user, passwd, host, port, dirs, timeout):
                               persistent=False)
 
 
+@functools.lru_cache()
+def _build_urlopener(ftp_tls=False, ssl_context=None, allow_insecure=False):
+    """
+    Helper for building a `urllib.request.build_opener` which handles TLS/SSL.
+    """
+
+    ssl_context = dict(it for it in ssl_context) if ssl_context else {}
+    cert_chain = {}
+    if 'certfile' in ssl_context:
+        cert_chain.update({
+            'certfile': ssl_context.pop('certfile'),
+            'keyfile': ssl_context.pop('keyfile', None),
+            'password': ssl_context.pop('password', None)
+        })
+    elif 'password' in ssl_context or 'keyfile' in ssl_context:
```

Comment:
The only way I can think of that would make sense to test this would be to start up a little SSL server, generate certificates, and check that a file can be downloaded from the server.  I could do that but it's probably more effort than it's worth (which to me also maybe suggests that the feature should be omitted entirely for now, though I think it's mostly harmless as well...